lindi/LindiH5pyFile/LindiH5pyDataset.py

```diff
@@ -0,0 +1,271 @@
+                )
+                return ret
+            else:
+                raise TypeError(
+                    f"Compound dataset {self.name} does not support selection with {selection}"
+                )
+
+        # We use zarr's slicing, except in the case of a scalar dataset
+        if self.ndim == 0:
+            # make sure selection is ()
+            if selection != ():
+                raise TypeError(f'Cannot slice a scalar dataset with {selection}')
+            return zarr_array[0]
+        return zarr_array[selection]
+
+    def _get_external_hdf5_client(self, url: str) -> h5py.File:
+        if url not in _external_hdf5_clients:
+            if url.startswith("http://") or url.startswith("https://"):
+                ff = remfile.File(url)
+            else:
+                ff = open(url, "rb")  # this never gets closed
+            _external_hdf5_clients[url] = h5py.File(ff, "r")
+        return _external_hdf5_clients[url]
+
+
+def _resolve_references(x: Any):
+    if isinstance(x, dict):
+        if '_REFERENCE' in x:
+            return LindiH5pyReference(x['_REFERENCE'])
+        else:
+            for k, v in x.items():
+                x[k] = _resolve_references(v)
+    elif isinstance(x, list):
+        for i, v in enumerate(x):
+            x[i] = _resolve_references(v)
+    elif isinstance(x, np.ndarray):
+        if x.dtype == object or x.dtype is None:
+            view_1d = x.reshape(-1)
+            for i in range(len(view_1d)):
+                view_1d[i] = _resolve_references(view_1d[i])
+    return x
+
+
+class LindiH5pyDatasetCompoundFieldSelection:
+    """
+    This class is returned when a compound dataset is indexed with a field name.
+    For example, if the dataset has dtype [('x', 'f4'), ('y', 'f4')], then we
+    can do dataset['x'][0] to get the first x value. The dataset['x'] returns an
+    object of this class.
+    """
+    def __init__(self, *, dataset: LindiH5pyDataset, ind: int, dtype: np.dtype):
+        self._dataset = dataset  # The parent dataset
+        self._ind = ind  # The index of the field in the compound dtype
+        self._dtype = dtype  # The dtype of the field
+        if self._dataset.ndim != 1:
+            # For now we only support 1D datasets
+            raise TypeError(
+                f"Compound field selection only implemented for 1D datasets, not {self._dataset.ndim}D"
+            )
+        if not isinstance(self._dataset._dataset_object, zarr.Array):
+            raise TypeError(
+                f"Compound field selection only implemented for zarr.Array, not {type(self._dataset._dataset_object)}"
+            )
+        za = self._dataset._dataset_object
+        self._zarr_array = za
+        # Prepare the data in memory
+        d = [za[i][self._ind] for i in range(len(za))]
+        if self._dtype == h5py.Reference:
+            # Convert to LindiH5pyReference
+            d = [LindiH5pyReference(x['_REFERENCE']) for x in d]
```

Comment:
```suggestion
            d = _resolve_references(d)
```

Reply:
But I think we'd want it to raise an exception if every value in d was not a valid reference.

Reply:
Got it. That makes sense.